src/core/desktopentry.cpp

```diff
@@ -407,7 +392,112 @@ DesktopEntry* DesktopEntryManager::heuristicLookup(const QString& name) {
 
 ObjectModel<DesktopEntry>* DesktopEntryManager::applications() { return &this->mApplications; }
 
-DesktopEntries::DesktopEntries() { DesktopEntryManager::instance(); }
+void DesktopEntryManager::handleFileChanges() {
+	qCDebug(logDesktopEntry) << "Directory change detected, performing full rescan";
+
+	if (this->scanInProgress) {
+		qCDebug(logDesktopEntry) << "Scan already in progress, skipping";
+		return;
+	}
+
+	this->scanInProgress = true;
+	auto* scanner = new DesktopEntryScanner(this);
+	QThreadPool::globalInstance()->start(scanner);
+}
+
+QString DesktopEntryManager::extractIdFromPath(const QString& path) {
+	auto info = QFileInfo(path);
+	auto id = info.completeBaseName();
+
+	auto appIdx = path.lastIndexOf("/applications/");
+	if (appIdx != -1) {
+		auto relativePath = path.mid(appIdx + 14);
+		auto relInfo = QFileInfo(relativePath);
+
+		auto dirPath = relInfo.path();
+		if (dirPath != ".") {
+			id = dirPath.replace('/', '-') + '-' + relInfo.completeBaseName();
+		}
+	}
+
+	return id;
+}
+
+void DesktopEntryManager::onScanCompleted(const DesktopEntryScanResults& scanResults) {
+	auto guard = qScopeGuard([this] { this->scanInProgress = false; });
+
+	auto oldEntries = this->desktopEntries;
+	auto newEntries = QHash<QString, DesktopEntry*>();
+	auto newLowercaseEntries = QHash<QString, DesktopEntry*>();
+
+	for (const auto& parsed: scanResults) {
+		auto* dentry = new DesktopEntry(parsed.id, this);
+		dentry->parseEntry(parsed.content);
+
+		if (!dentry->isValid()) {
+			qCDebug(logDesktopEntry) << "Skipping desktop entry" << parsed.id;
+			delete dentry;
+			continue;
+		}
+
+		qCDebug(logDesktopEntry) << "Found desktop entry" << parsed.id;
+
+		auto lowerId = parsed.id.toLower();
+
+		auto conflictingId = newEntries.contains(parsed.id);
+
+		if (conflictingId) {
+			qCDebug(logDesktopEntry) << "Replacing old entry for" << parsed.id;
+			delete newEntries.value(parsed.id);
+			newEntries.remove(parsed.id);
+			newLowercaseEntries.remove(lowerId);
+		}
+
+		newEntries.insert(parsed.id, dentry);
+
+		if (newLowercaseEntries.contains(lowerId)) {
+			qCInfo(logDesktopEntry).nospace()
+			    << "Multiple desktop entries have the same lowercased id " << lowerId
+			    << ". This can cause ambiguity when byId requests are not made with the correct case "
+			       "already.";
+
+			newLowercaseEntries.remove(lowerId);
+		}
+
+		newLowercaseEntries.insert(lowerId, dentry);
+	}
+
+	this->desktopEntries = newEntries;
+	this->lowercaseDesktopEntries = newLowercaseEntries;
+
+	auto newApplications = QVector<DesktopEntry*>();
+	for (auto& entry: this->desktopEntries.values()) {
+		if (!entry->noDisplay()) {
+			newApplications.append(entry);
+		}
+	}
+
+	this->mApplications.diffUpdate(newApplications);
+
+	for (auto* e: oldEntries) {
+		if (!this->desktopEntries.contains(e->mId)) {
+			e->deleteLater();
+		}
+	}
+	oldEntries.clear();
+
+	emit applicationsChanged();
+}
```

Comment:
DiffUpdate will not do anything useful when the values are different. This has to be handled manually. See above note

Reply:
Re-using existing entries* when IDs match now, deleting leftovers